File: packages/bare-host-protocol/src/request-tracker.ts
import {
  PROTOCOL_VERSION,
  type ErrorEnvelope,
  type HostMethod,
  type RequestEnvelope,
  type ResponseEnvelope,
} from "./messages.js";

export class RequestTracker {
  private nextId = 1;
  private readonly pending = new Set<number>();

  request(method: HostMethod): RequestEnvelope {
    if (!Number.isSafeInteger(this.nextId)) {
      throw new Error("control request id space is exhausted");
    }
    const id = this.nextId++;
    this.pending.add(id);
    return { version: PROTOCOL_VERSION, kind: "request", id, method };
  }

  accept(
    envelope: ResponseEnvelope | ErrorEnvelope,
  ): ResponseEnvelope | ErrorEnvelope {
    if (!this.pending.delete(envelope.id)) {
      throw new Error(`unknown response id: ${envelope.id}`);
    }
    return envelope;
  }
}

File: android/barekit-host/src/main/java/io/github/ttalab/barekit/host/protocol/RequestTracker.kt
package io.github.ttalab.barekit.host.protocol

import kotlinx.serialization.json.JsonElement

class RequestTracker {
  private var nextId = 1L
  private val pending = mutableSetOf<Long>()

  fun request(method: String, params: JsonElement? = null): RequestEnvelope {
    require(
      method == "ping" ||
        method == "status" ||
        method == "stop" ||
        method == "configure"
    ) {
      "unsupported control request method"
    }
    check(nextId <= 9_007_199_254_740_991) {
      "control request id space is exhausted"
    }
    val request = RequestEnvelope(PROTOCOL_VERSION, "request", nextId++, method, params)
    pending += request.id
    return request
  }

  fun accept(envelope: HostEnvelope): HostEnvelope {
    require(envelope is ResponseEnvelope || envelope is ErrorEnvelope) {
      "only response and error envelopes complete requests"
    }
    val id = when (envelope) {
      is ResponseEnvelope -> envelope.id
      is ErrorEnvelope -> envelope.id
      else -> error("unreachable")
    }
    require(pending.remove(id)) { "unknown response id: $id" }
    return envelope
  }
}
